src/lib.rs

```diff
@@ -7,6 +7,20 @@ use std::vec::IntoIter;
 #[cfg(feature = "serde")]
 use serde::{de::Error, Deserialize, Deserializer, Serialize, Serializer};
 
+/// Call this macro to assert that the current code path is unreachable due to
+/// a non-empty invariant. This avoids a lot of ceremony in the implementation,
+/// since almost all unsafe code in this crate relies on the same invariant.
+macro_rules! unreachable_non_empty {
+    () => {{
+        #[cfg(debug_assertions)]
+        ::std::unreachable!();
+        #[allow(unreachable_code)]
+        unsafe {
+            ::std::hint::unreachable_unchecked()
+        }
```

Comment:
I think the name `unreachable_non_empty` has imply it's only safe to use when non empty invariant is broken which won't happen in safe code, so it should be fine?

Reply:
To me `unreachable_*` sounds like it has similar semantics as the std `unreachable` macro, which implies it's totally "safe" to reach (would just panic) and not actually UB (which might result in miscompilations if reachable, which is be a lot worse).
The byteorder crate for example prefixes their macros with `unsafe_` [(ref)](https://github.com/BurntSushi/byteorder/commit/5d9d0386488d0763954f14a5cb02b1c0afccbdcb), but I'm not sure why they do that instead of just not using an unsafe block in the macro and requiring it at macro call site.

But I guess that's really nitpicky, a quick look at the macro definition reveals what it really is. Maybe it's fine

Reply:
these are good arguments. I originally did it this way because I didn't want to repeat the same exact `// SAFETY:` comment, but it makes sense that explicitness > terseness in this case.